src/Swarm/TUI/Controller.hs

```diff
@@ -450,21 +444,27 @@ validateREPLForm s =
   validate = setFieldValid (isRight result) REPLInput
 
 -- | Update our current position in the REPL history.
-adjReplHistIndex :: (Int -> Int -> Int) -> AppState -> AppState
-adjReplHistIndex (+/-) s =
-  s & uiState . uiReplHistIdx .~ newIndex
-    & (if curIndex == -1 then saveLastEntry else id)
-    & (if newIndex /= curIndex then uiState . uiReplForm %~ updateFormState newEntry else id)
+adjReplHistIndex :: TimeDir -> AppState -> AppState
+adjReplHistIndex d s =
+  ns
+    & (if s ^. replIx == s ^. replLen then saveLastEntry else id)
```

Comment:
The condition `s ^. replIx == s ^. replLen` expresses that the input buffer is the current focus of the `replHistory` belonging to the `uiState` of `s`. I can see that immediately now, because I know the indexing convention for `REPLHistory`. Should the controller as a client of `REPLHistory` really know and care about this convention (which incidentally was changed for this new implementation)? I can imagine myself wondering in six months what is going on here because I forgot that the REPL index being the REPL length means focus on the input buffer.

I therefore propose the following change:
 ```suggestion
    & (if replAtInputBuffer (s ^. repl) then saveLastEntry else id)
```
This uses a new (probably badly named) function 
```haskell
replAtInputBuffer :: REPLHistory -> Bool
replAtInputBuffer replHist = replHist ^. replIndex == replHist ^. to replLength
```
which in turn can be moved to `Model.hs`, which should be the right place for internals of `REPLHistory`.